staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -424,30 +424,70 @@ func (wc *watchChan) startWatching(watchClosedCh chan struct{}, initialEventsEnd
 	close(watchClosedCh)
 }
 
-// processEvent processes events from etcd watcher and sends results to resultChan.
-func (wc *watchChan) processEvent(wg *sync.WaitGroup) {
-	defer wg.Done()
+// processEvents processes events from etcd watcher and sends results to resultChan.
+func (wc *watchChan) processEvents(wg *sync.WaitGroup) {
+	processingQueue := make(chan chan *watch.Event, processEventConcurrency-1)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		wc.scheduleEventProcessing(wg, processingQueue)
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		wc.collectEventProcessing(processingQueue)
+	}()
+}
 
+func (wc *watchChan) scheduleEventProcessing(wg *sync.WaitGroup, processingQueue chan<- chan *watch.Event) {
```

Comment:
I thing this implementation is much easier to follow.

What I'm wondering is whether we can actually encapsulate the logic into some helper struct?

Reply:
What exact encapsulation you have in mind? Would like to align on what should be encapsulated. Should it be generic (transformation passed via generic func) or have pointer access to watchChan?

Reply:
Done, PTAL